keyboards/bastardkb/charybdis/charybdis.c

```diff
@@ -0,0 +1,245 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "charybdis.h"
+#include "transactions.h"
+#include <string.h>
+#include "drivers/sensors/pmw3360.h"
+#include "print.h"
+
+#ifndef CHARYBDIS_DPI_OPTIONS
+#    define CHARYBDIS_DPI_OPTIONS \
+        { 600, 800, 1000, 1200, 1400, 1600, 1800, 2000, 2200, 2400, 2600, 2800  }
+#endif
+#ifndef CHARYBDIS_DPI_DEFAULT
+#    define CHARYBDIS_DPI_DEFAULT 0
+#endif
+
+#ifndef CHARYBDIS_SNIPER_OPTIONS
+#define CHARYBDIS_SNIPER_OPTIONS \
+        { 100, 200, 300, 400 }
+#endif
+#ifndef CHARYBDIS_SNIPER_DEFAULT
+#    define CHARYBDIS_SNIPER_DEFAULT 0
+#endif
+#ifndef CHARYBDIS_AUTO_SNIPER_LAYER
+#    define CHARYBDIS_AUTO_SNIPER_LAYER 4
+#endif
+
+#ifndef CHARYBDIS_DRAGSCROLL_DPI
+#    define CHARYBDIS_DRAGSCROLL_DPI 100 // Fixed-DPI Drag Scroll
+#endif
+
+#ifndef CHARYBDIS_DRAGSCROLL_MULTIPLIER
+#    define CHARYBDIS_DRAGSCROLL_MULTIPLIER 0.2 // Variable-DPI Drag Scroll
+#endif
+
+
+#ifndef CHARYBDIS_TRACKBALL_SPEED_DIVIDER
+#    define CHARYBDIS_TRACKBALL_SPEED_DIVIDER 12 
+#endif
+
+keyboard_config_t keyboard_config;
+uint16_t          dpi_array[] = CHARYBDIS_DPI_OPTIONS;
+#define DPI_OPTION_SIZE (sizeof(dpi_array) / sizeof(uint16_t))
+
+uint16_t          sniper_array[] = CHARYBDIS_SNIPER_OPTIONS;
+#define SNIPER_OPTION_SIZE (sizeof(sniper_array) / sizeof(uint16_t))
+
+bool    is_drag_scroll    = false;
+
+scroll_inertia_t scroll_inertia; // Scroll value storage to make scrolling slower
+
+__attribute__((weak)) void process_mouse_user(report_mouse_t* mouse_report, int16_t x, int16_t y) {
+    mouse_report->x = x;
+    mouse_report->y = y;
+}
+
+__attribute__((weak)) void process_mouse(report_mouse_t* mouse_report) {
+    report_pmw_t data = pmw_read_burst();
+    if (data.isOnSurface && data.isMotion) {
+        // Wrap to HID size
+        data.dx = constrain(data.dx, -127, 127);
+        data.dy = constrain(data.dy, -127, 127);
+
+        mouse_report->x = -data.dx;
+        mouse_report->y = data.dy;
+    }
+}
+
+// on layer change, no matter where the change was initiated
+// Then runs keymap's layer change check
+__attribute__((weak)) layer_state_t layer_state_set_keymap(layer_state_t state) { return state; }
+layer_state_t                       layer_state_set_user(layer_state_t state) {
+
+    static bool is_sniper_on = false;
+    if (layer_state_cmp(state, CHARYBDIS_AUTO_SNIPER_LAYER) != is_sniper_on) {
+        is_sniper_on = layer_state_cmp(state, CHARYBDIS_AUTO_SNIPER_LAYER);
+        
+        if (is_sniper_on) {
+               pmw_set_cpi(sniper_array[keyboard_config.sniper_config]);
+           } else {
+               pmw_set_cpi(dpi_array[keyboard_config.dpi_config]);
+        } 
+    }
+    state = layer_state_set_keymap(state);
+    return state;
```

Comment:
```suggestion
    return state;
```

Reply:
so @drashna , I don't get the purpose of layer_state_set_keymap.
It seems to simply take a state in argument, and return it?

Then what's the point of calling it

Reply:
the `keymap` version is probably from my userspace.  which is because I use the userspace, and need an additional callback (kb -> userspace -> keymap). 

Here, you should use the kb version, and then call the user version.

Reply:
thanks. done.